src/profile-logic/line-timings.js

```diff
@@ -217,6 +217,219 @@ export function getStackLineInfoInverted(
+ * The following stacks all "collapse into" ("map to") call node 3:
+ * stack 3, 4, 6 and 7.
+ * Stack 8 maps to call node 4, which is a child of call node 3.
+ * All other stacks are outside the call path [A, B, C].
+ *
+ * In this function, we only compute "line hits" that are contributed to
+ * the given call node.
+ * Stacks 3, 4, 6 and 7 all contribute their time both as "self time"
+ * and as "total time" to call node 3, at the line numbers 30, 40, 60,
+ * and 70, respectively.
+ * Stack 8 also hits call node 3 at line 70, but does not contribute to
+ * call node 3's "self time", it only contributes to its "total time".
+ * All other stacks don't contribute to call node 3's self or total time.
+ *
+ * All stacks can contribute no more than one line in the given call node.
+ * This is different from the getStackLineInfo function above, where each
+ * stack can hit many lines in the same file, because all of the ancestor
+ * stacks are taken into account, rather than just one of them. Concretely,
+ * this means that in the returned StackLineInfo, each stackLines[stack]
+ * set will only contain at most one element.
+ *
+ * The returned StackLineInfo is computed as follows:
+ *   selfLine[stack]:
+ *     For stacks that map to the given call node, this is stack.frame.line.
+ *     For all other stacks this is null.
+ *   stackLines[stack]:
+ *     For stacks that map to the given call node or one of its descendent
+ *     call nodes, this is a set containing one element, which is
+ *     ancestorStack.frame.line, where ancestorStack maps to the given call
+ *     node.
+ *     For all other stacks, this is null.
+ */
+export function getStackLineInfoForCallNodeNonInverted(
+  stackTable: StackTable,
+  frameTable: FrameTable,
+  callNodeIndex: IndexIntoCallNodeTable,
+  { stackIndexToCallNodeIndex }: CallNodeInfo
+): StackLineInfo {
+  // "self line" == "the line which a stack's self time is contributed to"
+  const callNodeSelfLineForAllStacks = [];
+  // "total lines" == "the set of lines whose total time this stack contributes to"
+  const callNodeTotalLinesForAllStacks = [];
+
+  // This loop takes advantage of the fact that the stack table is topologically ordered:
+  // Prefix stacks are always visited before their descendants.
+  for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+    let selfLine: LineNumber | null = null;
+    let totalLines: Set<LineNumber> | null = null;
+
+    if (stackIndexToCallNodeIndex[stackIndex] === callNodeIndex) {
+      // This stack contributes to the call node's self time.
+      // We don't need to check the stack's func or file because it'll be
+      // the same as the given call node's func and file.
+      const frame = stackTable.frame[stackIndex];
+      selfLine = frameTable.line[frame];
+      if (selfLine !== null) {
+        totalLines = new Set([selfLine]);
+      }
+    } else {
+      // Inherit total time contribution from the parent stack. selfLine is null.
```

Comment:
Is that what makes stack 8 above contribute to call node 3's total time? It could help if you'd refer to the example in comments in the algorithm.

Reply:
Correct! Will do.